File: leetcode/1334.cpp
// 2 11m
class Solution {
public:
    int findTheCity(int n, vector<vector<int>>& edges, int maxD) {
        vector<vector<int>> dp(n, vector<int>(n, INT_MAX/2));
        
        for(int i = 0; i < n; i++) dp[i][i] = 0;
        for(auto &e: edges){
            int u = e[0], v = e[1], cost = e[2];
            dp[u][v] = dp[v][u] = cost;
        }
        
        for(int k = 0; k < n; k++){
            for(int i = 0; i < n; i++){
                for(int j = 0; j < n; j++)
                    dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j]); // bug1: INT overflow due to INT_MAX + INT_MAX
            }
        }
        
        int res = 0, mincnt = INT_MAX;
        for(int i = 0; i < n; i++){
            int cnt = 0;
            for(int j = 0; j < n; j++){
                if( i == j ) continue;
                if( dp[i][j] <= maxD )
                    cnt++;
            }
            
            if( cnt <= mincnt ){
                res = i;
                mincnt = cnt;
            }
        }
        
        return res;
    }
};

File: leetcode/1334.swift
// 1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance
class Solution {
    func findTheCity(_ n: Int, _ edges: [[Int]], _ distanceThreshold: Int) -> Int {
        var dp : [[Int]] = Array(repeating: Array(repeating: Int.max/2, count: n), count: n)
        
        for edge in edges {
            dp[edge[0]][edge[1]] = edge[2]
            dp[edge[1]][edge[0]] = edge[2]
        }
        
        for i in 0..<n {
            dp[i][i] = 0
        }
        
        for k in 0..<n {
            for i in 0..<n {
                for j in 0..<n {
                    dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j])
                }
            }
        }
        
        var neighborCount: [Int] = Array(repeating: 0, count: n)
        for i in 0..<n {
            for j in 0..<n{
                if dp[i][j] <= distanceThreshold {
                    neighborCount[i] += 1
                }
            }
        }
        
        var city = 0
        var min_count = neighborCount[city]
        for i in 1..<n {
            if neighborCount[i] <= min_count {
                city = i
                min_count = neighborCount[i]
            }
        }
        
        return city
    }
}
